bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java

```diff
@@ -1355,7 +1364,21 @@ int handleNewWindowRequested(long pView, long pArgs) {
 			args.Release();
 			inNewWindow = false;
 		}
-	});
+	};
+
+	// In case a new window is opened using evaluate with scripts like
+	// 'window.open()', there can be a deadlock if we execute the handler
+	// asynchronously. And in case a new browser is initialized in the same
+	// environment inside an OpenWindowListener of another browser, there can also
+	// be a deadlock. Hence, generally it should be run asynchronously. However, a
+	// combination of opening a new window from evaluate and starting a new instance
+	// of browser inside the OpenWindowListener should be avoided.
+	if (inEvaluate) {
+		openWindowHandler.run();
+	} else {
+		browser.getDisplay().asyncExec(openWindowHandler);
+	}
```

Comment:
I would move the code to a new private method and make use of the JavaDoc instead of the comment (`//`). That way you can use `<pre>`-blocks in order to add example (javascript) code, which is easier to understand than just words.

Something like:

```suggestion
	doHandleOpenWindow(openWindowHandler);
```

And you add the new method like this:

```java
/**
 * Explain the situation here and ...
 *
 * <pre>
 * // add some JavaScript code in here
 * evaluate(window.open)
 * </pre>
 *
 */
private void doHandleOpenWindow(Runnable r) {
	if (inEvaluate) {
		r.run();
	} else {
		browser.getDisplay().asyncExec(r);
	}
}
```

Reply:
To me the method name doesn't justify what it does. It seems like it can accept any runnable. The reason why I didn't move that piece of code was as follows:

1. It has only one caller
2. It is very specific to the openWindow handler.

Hence, even if we separate the method, we need a better name. e.g. doHandleRunnableInEvaluate - however, this whole logic is very specific to openWindow event. So in my opinion we do not need to create a new method here until we are reusing this pattern in multiple places.

Reply:
Ok, but at least add code snippets to your comment or rephrase it so it's a bit simpler to grasp the issue. Something like: _calling 'evaluate(window.open)' may cause deadlocks if it runs asynchronously and starting a new browser instance in the OpenWindowListener ..._ etc etc

Reply:
Done